core/src/server/rpc_module.rs

```diff
@@ -881,25 +882,40 @@ impl SubscriptionSink {
 	///            SubscriptionClosed::Failed(e) => {
 	///                sink.close(e);
 	///            }
-	///         };
+	///         }).await;
 	///     });
 	/// });
 	/// ```
-	pub async fn pipe_from_try_stream<S, T, E>(&mut self, mut stream: S) -> SubscriptionClosed
+	pub async fn pipe_from_try_stream<S, T, E, F>(mut self, mut stream: S, on_close: F)
 	where
 		S: TryStream<Ok = T, Error = E> + Unpin,
 		T: Serialize,
 		E: std::fmt::Display,
+		F: FnOnce(SubscriptionClosed, SubscriptionSink),
 	{
-		let close_notify = match self.close_notify.clone() {
+		let conn_closed = match self.close_notify.clone() {
 			Some(close_notify) => close_notify,
-			None => return SubscriptionClosed::RemotePeerAborted,
+			None => {
+				on_close(SubscriptionClosed::RemotePeerAborted, self);
```

Comment:
I don't really understand the benefit of having this `on_close` closure versus returning, say, `(SubscriptionClosed, Self)`?

And that makes me wonder; would this method take `&mut self` instead then, so it wouldn't need to return `Self` at all, just `SubscriptionClosed`?

And then I wonder, would it be nicer to remove the `SubscriptionClosed::Success` variant and actually return `Result<(), SubscriptionClosed>` rather than `SubscriptionClosed` (and maybe then rename `SubscriptionClosed` to something like `SubscriptionCloseError`?

Reply:
I had to double check, but what I mean is that the closure doesn't protect against it being reused? It might be a little safer if it was handed `&mut self` instead of `self` though, but the safest approach if people need access to some of the functionality from `self`, but not all of it, would be to wrap `self` in some type to make sure it can only be used in a certain way?

Reply:
Dumb question; why can't it be reused?

Reply:
The refactoring [here](https://github.com/paritytech/jsonrpsee/blob/na-close-tasks-when-unsubscribed/core/src/server/rpc_module.rs#L904) which takes the `unsubscribed oneshot` which is why I wanted the `pipe_from_sink(self)` rather than `pipe_from_sink(&mut self)`

So if I do this:

```rust
  sink.pipe_from(stream1).await;
  // this will compile but not work, unsubscribe oneshot is already consumed via `Option::take`
  sink.pipe_from(stream2).await;
```

But yeah I can do nasty things in the exit closure too you correct,

```rust
  sink.pipe_from(stream1, |_, _| { i can do nasty stuff here as well but async stuff right }).await;
  // this won't compile
  sink.pipe_from(stream2).await;
```

Reply:
reverted good catch James.